KnowledgeExtractor/WikipediaKnowledgeGraphExtractor.cs

```diff
@@ -96,34 +101,70 @@ public static KnowledgeGraph ExtractKnGraphFromUris(List<UriAndOriginalGraphType
         public static KnowledgeGraph GetKnowledgeGraphFromWikipedia()
         {
             KnowledgeGraph result = new KnowledgeGraph();
-
-            // get the kn graph from the uris
             result = ExtractKnGraphFromUris(WikipediaPagesToParse);
+            AddEdgesBetweenAlgorithmsAndDataStructures(ref result);
+            return result;
+        }
+
+        public static void AddEdgesBetweenAlgorithmsAndDataStructures(ref KnowledgeGraph graph)
+        {
+            List<WordCount> dataStructureWordsInGraph = GetDataStructureWordsInEachWikiPage();
+            Dictionary<string, KnGNode> dataStructureNodes = GetDataStructureNodesWithoutDuplicates(graph);
+            
+            foreach (var dataStructureWordsToAddInOneNode in dataStructureWordsInGraph)
+            {
+                AddDataStructureWordsToNode(graph: ref graph, wordsToAddToNode: dataStructureWordsToAddInOneNode, dataStructureNodes: dataStructureNodes);
+            }
+        }
 
-            // add the additional edges from the data structures word count to show what data structures each algorithm and data structure is using
-            // get the word counts from the json file
-            List<WordCount> dsWordsForGraph = JsonConvert.DeserializeObject<List<WordCount>>(File.ReadAllText("../../../dataStructureWordsCountForNodesInGraph.json"));
-            // get the nodes for the DS words
-            List<KnGNode> dataStructureNodes = result.KnGraph.Where(n => n.OriginalGraphType == OriginalGraphType.DataStructuresKnGraph).ToList();
-            List<string> dsWordsToIgnore = new List<string>() { "other", "union", "reference", "arrays", "image", "matrix", "integer", "character", "list", "stack", "queue", "record", "bitmap", "set" };
-            // add the additional edges to the graph according to the json file
-            foreach (var dsWordsForNode in dsWordsForGraph)
+        private static void AddDataStructureWordsToNode(ref KnowledgeGraph graph, WordCount wordsToAddToNode, Dictionary<string, KnGNode> dataStructureNodes)
+        {
+            List<string> words = GetWords(wordsToAddToNode);
+            List<string> filteredWords = GetFilteredWords(words, DataStructureWordsToIgnore);
+
+            foreach (var word in filteredWords)
             {
-                int nodeIndex = dsWordsForNode.Index;
-                List<string> dsWords = dsWordsForNode.WordsCount.Keys.ToList();
-                dsWords = dsWords.Select(s => s.ToLowerInvariant()).ToList();
-                List<string> dsWordsFiltered = dsWords.Except(dsWordsToIgnore).ToList();
+                AddEdgeToKnowledgeGraph(graph: ref graph, nodeIndex: wordsToAddToNode.Index, nodeToAdd: dataStructureNodes[word]);
```

Comment:
Again I would probably not have explicitly named the parameters. I think their meaning is clear enough without the names

Reply:
I would keep this one as is because we wouldn't know that dataStructureNodes[word] is the node to add. And my goal with cleaning the code is to get it to the state where if I forget about it and its context, I can come and read it without tiring and without getting frustrated. So I want to context to be obvious very quickly.